base/linalg/bidiag.jl

```diff
@@ -224,8 +224,163 @@ end
 /(A::Bidiagonal, B::Number) = Bidiagonal(A.dv/B, A.ev/B, A.isupper)
 ==(A::Bidiagonal, B::Bidiagonal) = (A.dv==B.dv) && (A.ev==B.ev) && (A.isupper==B.isupper)
 
-SpecialMatrix = Union{Bidiagonal, SymTridiagonal, Tridiagonal, AbstractTriangular}
-*(A::SpecialMatrix, B::SpecialMatrix)=full(A)*full(B)
+
+
+BiTriSym = Union{Bidiagonal, Tridiagonal, SymTridiagonal}
+BiTri = Union{Bidiagonal, Tridiagonal}
+
+
+A_mul_B!(C::AbstractMatrix, A::SymTridiagonal, B::BiTriSym) = A_mul_B_td!(C, A, B)
+A_mul_B!(C::AbstractMatrix, A::BiTri, B::BiTriSym) = A_mul_B_td!(C, A, B)
+A_mul_B!(C::AbstractMatrix, A::BiTriSym, B::BiTriSym) = A_mul_B_td!(C, A, B)
+A_mul_B!(C::AbstractMatrix, A::AbstractTriangular, B::BiTriSym) = A_mul_B_td!(C, A, B)
+A_mul_B!(C::AbstractMatrix, A::AbstractMatrix, B::BiTriSym) = A_mul_B_td!(C, A, B)
+A_mul_B!(C::AbstractVector, A::BiTri, B::AbstractVector) = A_mul_B_td!(C, A, B)
+A_mul_B!(C::AbstractMatrix, A::BiTri, B::AbstractVecOrMat) = A_mul_B_td!(C, A, B)
+A_mul_B!(C::AbstractVecOrMat, A::BiTri, B::AbstractVecOrMat) = A_mul_B_td!(C, A, B)
+
+
+
+function check_A_mul_B!_sizes(C, A, B)
+    nA, mA = size(A)
+    nB, mB = size(B)
+    nC, mC = size(C)
+    if !(nA == nC)
+        throw(DimensionMismatch("Sizes size(A)=$(size(A)) and size(C) = $(size(C)) must match at first entry."))
+    elseif !(mA == nB)
+        throw(DimensionMismatch("Second entry of size(A)=$(size(A)) and first entry of size(B) = $(size(B)) must match."))
+    elseif !(mB == mC)
+        throw(DimensionMismatch("Sizes size(B)=$(size(B)) and size(C) = $(size(C)) must match at first second entry."))
+    end
+end
+
+function A_mul_B_td!(C::AbstractMatrix, A::BiTriSym, B::BiTriSym)
+
+    check_A_mul_B!_sizes(C, A, B)
+    n = size(A,1)
+    n <= 3 && return A_mul_B!(C, full(A), full(B))
+
+    fill!(C, zero(eltype(C)))
```

Comment:
Actually I did not even add that particular method, it[ existed before](https://github.com/JuliaLang/julia/blob/acaf9aaffe3ff8e55dcd4824cfcfcc7a784e0244/base/linalg/triangular.jl#L50). I added analogues for Bidiagonal etc.

Initially I called this kind of function `fillslots!`, until I discovered that it already existed in the Triangular case under the name fill!. It would probably be saver if fill! would always commute with full or throw an error if this is not possible.
```full(fill!(A, a)) ==  fill!(full(A), a)```
Anyway I don't have a strong opinion about renaming fill! in cases where it does not commute with full and certainly did not want to break existing behaviour.